File: src/org/mtransit/parser/mt/data/MAgency.java
package org.mtransit.parser.mt.data;

import org.apache.commons.lang3.StringUtils;
import org.mtransit.parser.Constants;
import org.mtransit.parser.gtfs.data.GRouteType;

public class MAgency implements Comparable<MAgency> {

	public static final int ROUTE_TYPE_LIGHT_RAIL = GRouteType.LIGHT_RAIL.getId();
	public static final int ROUTE_TYPE_SUBWAY = GRouteType.SUBWAY.getId();
	public static final int ROUTE_TYPE_TRAIN = GRouteType.TRAIN.getId();
	public static final int ROUTE_TYPE_BUS = GRouteType.BUS.getId();
	public static final int ROUTE_TYPE_FERRY = GRouteType.FERRY.getId();

	private String id;
	private String timezone;
	private String color;
	private Integer type;

	public MAgency(String id, String timezone, String color, Integer type) {
		this.id = id;
		this.timezone = timezone;
		this.color = color;
		this.type = type;
	}

	@Override
	public int hashCode() {
		return (this.id == null ? 0 : this.id.hashCode()) + (this.timezone == null ? 0 : this.timezone.hashCode())
				+ (this.color == null ? 0 : this.color.hashCode()) + (this.type == null ? 0 : this.type.hashCode());
	}

	@Override
	public String toString() {
		return new StringBuilder() //
				.append(this.id) //
				.append(Constants.COLUMN_SEPARATOR) //
				.append(this.timezone) //
				.append(Constants.COLUMN_SEPARATOR) //
				.append(this.color) //
				.append(Constants.COLUMN_SEPARATOR) //
				.append(this.type) //
				.toString();

	}

	public String getId() {
		return id;
	}

	public String getTimezone() {
		return timezone;
	}

	public String getColor() {
		return color;
	}

	public Integer getType() {
		return type;
	}

	@Override
	public int compareTo(MAgency otherAgency) {
		if (otherAgency == null) {
			return +1;
		}
		if (this.id != null) {
			if (this.id.equals(otherAgency.id)) {
				return this.id.compareTo(otherAgency.id);
			}
		} else if (otherAgency.id != null) {
			return -1;
		}
		if (this.timezone != null) {
			if (this.timezone.equals(otherAgency.timezone)) {
				return this.timezone.compareTo(otherAgency.timezone);
			}
		} else if (otherAgency.timezone != null) {
			return -1;
		}
		if (this.color != null) {
			if (this.color.equals(otherAgency.color)) {
				return this.color.compareTo(otherAgency.color);
			}
		} else if (otherAgency.color != null) {
			return -1;
		}
		if (this.type != null) {
			if (this.type.equals(otherAgency.type)) {
				return this.type.compareTo(otherAgency.type);
			}
		} else if (otherAgency.type != null) {
			return -1;
		}
		return 0;
	}

	@Override
	public boolean equals(Object obj) {
		MAgency o = (MAgency) obj;
		if (!StringUtils.equals(this.id, o.id)) {
			return false;
		}
		if (!StringUtils.equals(this.timezone, o.timezone)) {
			return false;
		}
		if (!StringUtils.equals(this.color, o.color)) {
			return false;
		}
		if (this.type != null) {
			if (!this.type.equals(o.type)) {
				return false;
			}
		} else if (o.type != null) {
			return false;
		}
		return true;
	}
}

File: src/main/java/org/mtransit/parser/mt/data/MAgency.kt
package org.mtransit.parser.mt.data

import org.mtransit.parser.Constants
import org.mtransit.parser.gtfs.GAgencyTools
import org.mtransit.parser.gtfs.data.GAgency
import org.mtransit.parser.gtfs.data.GIDs
import org.mtransit.parser.gtfs.data.GRouteType
import org.mtransit.parser.gtfs.data.GSpec

data class MAgency(
    val idInt: Int,
    val name: String,
    val url: String,
    val timezone: String,
    val color: String,
    val type: Int
) : Comparable<MAgency?> {

    constructor(
        gAgency: GAgency,
        agencyTools: GAgencyTools,
        gSpec: GSpec,
    ) : this(
        gAgency.agencyIdInt,
        gAgency.agencyName,
        gAgency.agencyUrl,
        gAgency.agencyTimezone,
        agencyTools.getAgencyColor(gAgency, gSpec),
        agencyTools.agencyRouteType
    )

    @Deprecated(message = "Not memory efficient")
    @Suppress("unused")
    val id = _id

    private val _id: String
        get() {
            return GIDs.getString(idInt)
        }

    fun toFile() = buildString {
        append(_id) //
        append(Constants.COLUMN_SEPARATOR) //
        append(timezone) //
        append(Constants.COLUMN_SEPARATOR) //
        append(color) //
        append(Constants.COLUMN_SEPARATOR) //
        append(type) //
    }

    fun toStringPlus(): String {
        return toString() +
                "+(_id:$_id)"
    }

    override fun compareTo(other: MAgency?): Int {
        return when {
            other !is MAgency -> +1
            this.idInt != other.idInt -> this._id.compareTo(other._id)
            this.name != other.name -> this.name.compareTo(other.name)
            this.url != other.url -> this.url.compareTo(other.url)
            this.timezone != other.timezone -> this.timezone.compareTo(other.timezone)
            this.color != other.color -> this.color.compareTo(other.color)
            this.type != other.type -> this.type.compareTo(other.type)
            else -> 0
        }
    }

    companion object {
        @Suppress("unused")
        @JvmField
        val ROUTE_TYPE_LIGHT_RAIL = GRouteType.LIGHT_RAIL.id

        @Suppress("unused")
        @JvmField
        val ROUTE_TYPE_SUBWAY = GRouteType.SUBWAY.id

        @Suppress("unused")
        @JvmField
        val ROUTE_TYPE_TRAIN = GRouteType.TRAIN.id

        @Suppress("unused")
        @JvmField
        val ROUTE_TYPE_BUS = GRouteType.BUS.id

        @Suppress("unused")
        @JvmField
        val ROUTE_TYPE_FERRY = GRouteType.FERRY.id

        private const val ROUTE_COLOR_IS_AGENCY_COLOR_MIN_PCT = 0.66f

        @JvmStatic
        fun pickColorFromRoutes(gAgency: GAgency, gSpec: GSpec): String? {
            val agencyRoutes = gSpec.allRoutes?.filter { route -> !route.isDifferentAgency(gAgency.agencyIdInt) } ?: return null
            val otherAgencyRoutes = gSpec.getOtherRoutes(gAgency.agencyIdInt) ?: emptyList() // same agency, different type
            val allRoutes = agencyRoutes + otherAgencyRoutes
            allRoutes
                .mapNotNull { it.routeColor?.uppercase() }
                .filter { it.isNotBlank() }
                .groupingBy { it }
                .eachCount()
                .toList()
                .maxByOrNull { it.second }
                ?.let { (color, count) ->
                    if (count > (allRoutes.size * ROUTE_COLOR_IS_AGENCY_COLOR_MIN_PCT)) {
                        return color
                    }
                }
            return null
        }
    }
}